### PR TITLE
chore(stdlib): Add Graindoc to Set library

### PR DESCRIPTION
**Pull request by spotandjake**
chore(stdlib): Add GrainDoc To Set Library

related tp #702.

### CODE REVIEW

File: stdlib/set.gr [hunk 1]
```diff
@@ -15,14 +17,24 @@ record Set<k> {
 }
 
 // TODO: This could take an `eq` function to custom comparisons
+/**
+ * Creates a new, empty set with an initial storage size for the given number of elements.
```
Comment (marcusroberts) [inline stdlib/set.gr:21]: I’d add that this is just a best guess and doesn’t limit the size of the set which will grow anyway. 

Do we have any recommendations for making a good guess?

File: stdlib/set.gr [hunk 2]
```diff
@@ -140,6 +165,12 @@ let rec removeInBucket = (key, node) => {
   }
 }
 
+/**
+ * Removes the given value from the set.
+ *
```
Comment (marcusroberts) [inline stdlib/set.gr:170]: What happens if the value isn’t in the set?

File: stdlib/set.gr [hunk 3]
```diff
@@ -160,14 +191,31 @@ export let remove = (key, set) => {
...
+ * @param set: The set to determine the length from
+ * @returns The length of the set
+ */
 export let size = (set) => {
   set.size
 }
 
+/**
+ * Returns a boolean indicating weather or not the set is empty.
```
Comment (marcusroberts) [inline stdlib/set.gr:205]: Whether

File: stdlib/set.gr [hunk 4]
```diff
@@ -276,6 +388,13 @@ export let diff = (set1, set2) => {
   set
 }
 
+/**
+ * Creates a set from the intersection of the given sets.
+ *
+ * @param set1: The first set to merge
```
Comment (marcusroberts) [inline stdlib/set.gr:394]: I’m not sure I would use merge here as that suggests union to me

File: stdlib/set.gr [hunk 5]
```diff
@@ -210,6 +294,14 @@ export let reduce = (fn, init, set) => {
   acc
 }
 
+/**
+ * Keeps all values that the predicate returned `true` for from the set.
+ *
+ * @param fn: The function used to filter the set returning `true` indicates to keep the value
```
Comment (phated) [inline stdlib/set.gr:300]:
```suggestion
 * @param fn: The predicate function to indicate which elements to remove from the set, where returning `false` indicates the value should be removed
```

File: stdlib/set.gr [hunk 6]
```diff
@@ -210,6 +294,14 @@ export let reduce = (fn, init, set) => {
...
+ * @param set: The set to filter
```
Comment (phated) [inline stdlib/set.gr:301]:
```suggestion
 * @param set: The set to iterate
```

File: stdlib/set.gr [hunk 7]
```diff
@@ -222,14 +314,38 @@ export let filter = (predicate, set) => {
   }, keysToRemove);
 }
 
+/**
+ * Removes all values that the predicate returned true for from the set.
```
Comment (phated) [inline stdlib/set.gr:318]:
```suggestion
 * Removes elements from a set where a predicate function returns `true`.
```

File: stdlib/set.gr [hunk 8]
```diff
@@ -222,14 +314,38 @@ export let filter = (predicate, set) => {
...
+ *
+ * @param fn: The function used to filter the set returning `true` indicates to remove the value
```
Comment (phated) [inline stdlib/set.gr:320]:
```suggestion
 * @param fn: The predicate function to indicate which elements to remove from the set, where returning `true` indicates the value should be removed
```

File: stdlib/set.gr [hunk 9]
```diff
@@ -222,14 +314,38 @@ export let filter = (predicate, set) => {
...
+ * @param set: The set to filter
```
Comment (phated) [inline stdlib/set.gr:321]:
```suggestion
 * @param set: The set to iterate
```

File: stdlib/set.gr [hunk 10]
```diff
@@ -222,14 +314,38 @@ export let filter = (predicate, set) => {
...
+ * 
+ * @since 0.3.0
+ */
 export let reject = (predicate, set) => {
   filter((key) => !predicate(key), set)
 }
 
+/**
+ * Returns a list from the values of a set.
```
Comment (phated) [inline stdlib/set.gr:330]:
```suggestion
 * Converts a set into a list of its elements.
```

File: stdlib/set.gr [hunk 11]
```diff
@@ -222,14 +314,38 @@ export let filter = (predicate, set) => {
...
+ *
+ * @param set: The set to convert
+ * @returns The generated List
```
Comment (phated) [inline stdlib/set.gr:333]:
```suggestion
 * @returns A list containing all set values
```

File: stdlib/set.gr [hunk 12]
```diff
@@ -222,14 +314,38 @@ export let filter = (predicate, set) => {
...
+ */
 export let toList = (set) => {
   reduce((list, key) => [key, ...list], [], set)
 }
 
+/**
+ * Creates a set from a list.
+ *
+ * @param list: The list to convert
+ * @returns The generated set
```
Comment (phated) [inline stdlib/set.gr:345]:
```suggestion
 * @returns A set containing all list values
```

File: stdlib/set.gr [hunk 13]
```diff
@@ -238,10 +354,26 @@ export let fromList = (list) => {
   set
 }
 
+/**
+ * Returns an array from the values of a set.
```
Comment (phated) [inline stdlib/set.gr:358]:
```suggestion
 * Converts a set into an array of its elements.
```

File: stdlib/set.gr [hunk 14]
```diff
@@ -238,10 +354,26 @@ export let fromList = (list) => {
...
+ *
+ * @param set: The set to convert
+ * @returns The generated array
```
Comment (phated) [inline stdlib/set.gr:361]:
```suggestion
 * @returns An array containing all set values
```

File: stdlib/set.gr [hunk 15]
```diff
@@ -238,10 +354,26 @@ export let fromList = (list) => {
...
+ */
 export let toArray = (set) => {
   Array.fromList(toList(set))
 }
 
+/**
+ * Creates a set from an array.
+ *
+ * @param array: The array to convert
+ * @returns The generated set
```
Comment (phated) [inline stdlib/set.gr:373]:
```suggestion
 * @returns A set containing all array values
```

File: stdlib/set.gr [hunk 16]
```diff
@@ -250,6 +382,15 @@ export let fromArray = (array) => {
   set
 }
 
+/**
+ * Creates a single set from the union of the given sets.
+ *
+ * @param set1: The first set to union
+ * @param set2: The second set to union
+ * @returns The generated set
```
Comment (phated) [inline stdlib/set.gr:390]:
```suggestion
 * @returns A set containing all items of both sets
```

File: stdlib/set.gr [hunk 17]
```diff
@@ -250,6 +382,15 @@ export let fromArray = (array) => {
   set
 }
 
+/**
+ * Creates a single set from the union of the given sets.
```
Comment (phated) [inline stdlib/set.gr:386]:
```suggestion
 * Combines two sets into a single set containing all values from both sets.
```

File: stdlib/set.gr [hunk 18]
```diff
@@ -250,6 +382,15 @@ export let fromArray = (array) => {
...
+ *
+ * @param set1: The first set to union
```
Comment (phated) [inline stdlib/set.gr:388]:
```suggestion
 * @param set1: The first set to combine
```

File: stdlib/set.gr [hunk 19]
```diff
@@ -250,6 +382,15 @@ export let fromArray = (array) => {
...
+ * @param set2: The second set to union
```
Comment (phated) [inline stdlib/set.gr:389]:
```suggestion
 * @param set2: The second set to combine
```

File: stdlib/set.gr [hunk 20]
```diff
@@ -261,6 +402,15 @@ export let union = (set1, set2) => {
   set
 }
 
+/**
+ * Creates a set from the difference of the given sets.
```
Comment (phated) [inline stdlib/set.gr:406]:
```suggestion
 * Combines two sets into a single set containing only the values not shared between both sets.
```

File: stdlib/set.gr [hunk 21]
```diff
@@ -261,6 +402,15 @@ export let union = (set1, set2) => {
...
+ *
+ * @param set1: The first set to merge
+ * @param set2: The second set to merge
+ * @returns The generated set
```
Comment (phated) [inline stdlib/set.gr:410]:
```suggestion
 * @param set1: The first set to combine
 * @param set2: The second set to combine
 * @returns A set containing only unshared values from both sets
```

File: stdlib/set.gr [hunk 22]
```diff
@@ -276,6 +426,15 @@ export let diff = (set1, set2) => {
   set
 }
 
+/**
+ * Creates a new set from the values that are in both set1 and set2.
```
Comment (phated) [inline stdlib/set.gr:430]:
```suggestion
 * Combines two sets into a single set containing only the values shared between both sets.
```

File: stdlib/set.gr [hunk 23]
```diff
@@ -276,6 +426,15 @@ export let diff = (set1, set2) => {
...
+ *
+ * @param set1: The first set to intersect
+ * @param set2: The second set to intersect
+ * @returns A set containing the values that are in both sets
```
Comment (phated) [inline stdlib/set.gr:434]:
```suggestion
 * @param set1: The first set to combine
 * @param set2: The second set to combine
 * @returns A set containing only shared values from both sets
```

File: stdlib/set.gr [hunk 24]
```diff
@@ -293,6 +452,14 @@ export let intersect = (set1, set2) => {
 
 // TODO: Should return a Record type instead of a Tuple
 // Waiting on https://github.com/grain-lang/grain/issues/190
+/**
+ * Returns a tuple representing the internal state state of the set.
```
Comment (phated) [inline stdlib/set.gr:456]:
```suggestion
 * Provides data representing the internal state state of the set.
```

File: stdlib/set.gr [hunk 25]
```diff
@@ -293,6 +452,14 @@ export let intersect = (set1, set2) => {
...
+ *
+ * @param set: The set to operate on
```
Comment (phated) [inline stdlib/set.gr:458]:
```suggestion
 * @param set: The set to inspect
```

File: stdlib/set.gr [hunk 26]
```diff
@@ -293,6 +452,14 @@ export let intersect = (set1, set2) => {
...
+ * @returns The internal state of the set represented as a tuple
```
Comment (phated) [inline stdlib/set.gr:459]:
```suggestion
 * @returns The internal state of the set
```

**Review by phated — CHANGES_REQUESTED** [review]
These are looking a ton better. I made some suggestions, but I'm also trying to decide where we should be using `value` or `element` or `item` to be consistent.

**Comment by spotandjake** [comment]
The Changes Were Made.

File: stdlib/set.gr [hunk 27]
```diff
@@ -15,14 +19,28 @@ record Set<k> {
 }
 
 // TODO: This could take an `eq` function to custom comparisons
-export let makeSized = (size) => {
-  let buckets = Array.make(size, None);
+/**
+ * Creates a new empty Set with an initial storage of the given length. As items are added to the Set, the length may grow or shrink. Generally, you won't need to care about the length of your set and can use `Set.make()` instead.
```
Comment (phated) [inline stdlib/set.gr:23]:
```suggestion
 * Creates a new empty set with an initial storage of the given length. As values are added or removed, the length may grow or shrink. Generally, you won't need to care about the length of your set and can use `Set.make()` instead.
```

File: stdlib/set.gr [hunk 28]
```diff
@@ -116,6 +142,15 @@ export let add = (key, set) => {
   }
 }
 
+/**
+ * Determines if the set contains the given value.
+ *
+ * @param key: The key to search for
```
Comment (phated) [inline stdlib/set.gr:148]:
```suggestion
 * @param key: The value to search for
```

File: stdlib/set.gr [hunk 29]
```diff
@@ -116,6 +142,15 @@ export let add = (key, set) => {
...
+ * @param set: The set to be searched
```
Comment (phated) [inline stdlib/set.gr:149]:
```suggestion
 * @param set: The set to search
```

File: stdlib/set.gr [hunk 30]
```diff
@@ -160,14 +203,37 @@ export let remove = (key, set) => {
   }
 }
 
+/**
+ * Returns the number of values within the set.
+ *
+ * @param set: The set to inspect
+ * @returns The number of items in the set
```
Comment (phated) [inline stdlib/set.gr:210]:
```suggestion
 * @returns The number of elements in the set
```

File: stdlib/set.gr [hunk 31]
```diff
@@ -160,14 +203,37 @@ export let remove = (key, set) => {
...
+ * 
+ * @since 0.3.0
+ */
 export let size = (set) => {
   set.size
 }
 
+/**
+ * Determines if the set contains no items.
```
Comment (phated) [inline stdlib/set.gr:219]:
```suggestion
 * Determines if the set contains no elements.
```

File: stdlib/set.gr [hunk 32]
```diff
@@ -92,6 +110,14 @@ let rec nodeInBucket = (key, node) => {
   }
 }
 
+/**
+ * Adds a new value to the set. If the value already exists, nothing happens.
+ *
+ * @param key: The value to add to the set
```
Comment (phated) [inline stdlib/set.gr:116]:
```suggestion
 * @param key: The value to add
```

File: stdlib/set.gr [hunk 33]
```diff
@@ -140,6 +175,14 @@ let rec removeInBucket = (key, node) => {
   }
 }
 
+/**
+ * Removes the given value from the set. If the value does not exist, nothing happens.
+ *
+ * @param key: The value to remove from the set
```
Comment (phated) [inline stdlib/set.gr:181]:
```suggestion
 * @param key: The value to remove
```

File: stdlib/set.gr [hunk 34]
```diff
@@ -250,6 +382,15 @@ export let fromArray = (array) => {
   set
 }
 
+/**
+ * Combines two sets into a single set containing all values from both sets.
```
Comment (phated) [inline stdlib/set.gr:386]:
```suggestion
 * Combines two sets into a single set containing all elements from both sets.
```

File: stdlib/set.gr [hunk 35]
```diff
@@ -250,6 +382,15 @@ export let fromArray = (array) => {
...
+ *
+ * @param set1: The first set to combine
+ * @param set2: The second set to combine
+ * @returns A set containing all items of both sets
```
Comment (phated) [inline stdlib/set.gr:390]:
```suggestion
 * @returns A set containing all elements of both sets
```

File: stdlib/set.gr [hunk 36]
```diff
@@ -261,6 +402,15 @@ export let union = (set1, set2) => {
   set
 }
 
+/**
+ * Combines two sets into a single set containing only the values not shared between both sets.
```
Comment (phated) [inline stdlib/set.gr:406]:
```suggestion
 * Combines two sets into a single set containing only the elements not shared between both sets.
```

File: stdlib/set.gr [hunk 37]
```diff
@@ -261,6 +402,15 @@ export let union = (set1, set2) => {
...
+ *
+ * @param set1: The first set to combine
+ * @param set2: The second set to combine
+ * @returns A set containing only unshared values from both sets
```
Comment (phated) [inline stdlib/set.gr:410]:
```suggestion
 * @returns A set containing only unshared elements from both sets
```

File: stdlib/set.gr [hunk 38]
```diff
@@ -276,6 +426,15 @@ export let diff = (set1, set2) => {
   set
 }
 
+/**
+ * Combines two sets into a single set containing only the values shared between both sets.
```
Comment (phated) [inline stdlib/set.gr:430]:
```suggestion
 * Combines two sets into a single set containing only the elements shared between both sets.
```

File: stdlib/set.gr [hunk 39]
```diff
@@ -276,6 +426,15 @@ export let diff = (set1, set2) => {
...
+ *
+ * @param set1: The first set to combine
+ * @param set2: The second set to combine
+ * @returns A set containing only shared values from both sets
```
Comment (phated) [inline stdlib/set.gr:434]:
```suggestion
 * @returns A set containing only shared elements from both sets
```

**Review by phated — CHANGES_REQUESTED** [review]
A few more. I tried to consolidate "values" vs "elements" vs "items" by dropping "items" completely and using "values" for things going into or out-of a set and "elements" for things inside a set.

**Comment by spotandjake** [comment]
Changes Once Again Made

File: stdlib/set.gr [hunk 40]
```diff
@@ -140,6 +175,14 @@ let rec removeInBucket = (key, node) => {
   }
 }
 
+/**
+ * Removes the given value from the set. If the value does not exist, nothing happens.
```
Comment (ospencer) [inline stdlib/set.gr:179]:
```suggestion
 * Removes the given value from the set. If the value doesn't exist, nothing happens.
```

**Review by ospencer — COMMENTED** [review]
This looks fantastic! One tiny suggestion but otherwise thanks @spotandjake and @phated! Amazing work!

**Comment by ospencer** [comment]
Oh sorry I had some old pending stuff that you can ignore. Boo GitHub.

**Review by ospencer — APPROVED** [review]

**Review by phated — APPROVED** [review]
Absolute champion! Thanks 🥇

**Pull request merged**